File: product-processor-service/src/main/java/com/example/product_processor_service/configuration/kafka/factory/consumer/KafkaConsumerFactoryConfiguration.java
package com.example.product_processor_service.configuration.kafka.factory.consumer;

import com.example.product_processor_service.configuration.kafka.environment.KafkaEnvironment;
import com.example.product_processor_service.model.product.ProductPublisherDTO;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.apache.kafka.clients.consumer.ConsumerConfig;
import org.apache.kafka.common.serialization.Deserializer;
import org.apache.kafka.common.serialization.StringDeserializer;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.kafka.core.ConsumerFactory;
import org.springframework.kafka.core.DefaultKafkaConsumerFactory;
import org.springframework.kafka.support.serializer.JsonDeserializer;

import java.io.Serializable;
import java.util.HashMap;
import java.util.Map;

@Configuration
public class KafkaConsumerFactoryConfiguration {
    private final KafkaEnvironment kafkaEnvironment;

    public KafkaConsumerFactoryConfiguration(KafkaEnvironment kafkaEnvironment) {
        this.kafkaEnvironment = kafkaEnvironment;
    }

    @Bean
    public ObjectMapper objectMapper() {
        return new ObjectMapper();
    }

    @Bean
    public ConsumerFactory<String, ProductPublisherDTO> transactionConsumerFactory(ObjectMapper objectMapper) {
        var kafkaConsumerFactory = new DefaultKafkaConsumerFactory<String, ProductPublisherDTO>(buildKafkaConsumerProperties("com.example.product_manager_service.model.product.dto.ProductPublisherDto", "com.example.product_processor_service.model.product.ProductPublisherDTO", kafkaEnvironment.getKAFKA_GROUP_ID()));
        kafkaConsumerFactory.setKeyDeserializer(new StringDeserializer());
        kafkaConsumerFactory.setValueDeserializer(new JsonDeserializer<>(objectMapper));

        return kafkaConsumerFactory;
    }

    @Bean
    public ConsumerFactory<String, String> stringConsumerFactory(ObjectMapper objectMapper) {
        var kafkaConsumerFactory = new DefaultKafkaConsumerFactory<String, String>(buildKafkaConsumerProperties("com.example.product_manager_service.model.product.dto.ProductPublisherDto", "com.example.product_processor_service.model.product.ProductPublisherDTO", kafkaEnvironment.getKAFKA_GROUP_ID()));
        kafkaConsumerFactory.setKeyDeserializer(new StringDeserializer());
        kafkaConsumerFactory.setValueDeserializer(new StringDeserializer());

        return kafkaConsumerFactory;
    }

    private Map<String, Object> buildKafkaConsumerProperties(String classPathFrom, String classPathTo, String groupId){
        Map<String, Object> properties = new HashMap<>();
        properties.put(ConsumerConfig.GROUP_ID_CONFIG, groupId);
        properties.put(JsonDeserializer.TYPE_MAPPINGS, String.format("%s:%s", classPathFrom, classPathTo));
        properties.put(JsonDeserializer.TRUSTED_PACKAGES, "com.example.transaction_manager_service.*");
        properties.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, kafkaEnvironment.getKAFKA_BOOTSTRAP_SERVERS());
        properties.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");

        return properties;
    }
}

File: email-sender-service/src/main/kotlin/com/example/email_sender_service/configuration/kafka/factory/consumer/KafkaConsumerFactoryConfiguration.kt
package com.example.email_sender_service.configuration.kafka.factory.consumer

import com.example.email_sender_service.configuration.kafka.environment.KafkaEnvironment
import com.example.email_sender_service.model.mail.dto.MailMessage
import com.fasterxml.jackson.databind.ObjectMapper
import org.apache.kafka.clients.consumer.ConsumerConfig
import org.apache.kafka.common.serialization.StringDeserializer
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.kafka.core.ConsumerFactory
import org.springframework.kafka.core.DefaultKafkaConsumerFactory
import org.springframework.kafka.support.serializer.JsonDeserializer

@Configuration
class KafkaConsumerFactoryConfiguration(private val kafkaEnvironment: KafkaEnvironment) {
    @Bean
    fun objectMapper() = ObjectMapper()

    @Bean
    fun transactionConsumerFactory(objectMapper: ObjectMapper): ConsumerFactory<String, MailMessage> {
        val kafkaConsumerFactory: DefaultKafkaConsumerFactory<String, MailMessage> =
            DefaultKafkaConsumerFactory<String, MailMessage>(
                buildKafkaConsumerProperties(
                    "com.example.product_processor_service.model.mail.dto.MailMessage",
                    "com.example.email_sender_service.model.mail.dto.MailMessage",
                    kafkaEnvironment.KAFKA_TOPIC_USER_GROUP_ID
                )
            )
        kafkaConsumerFactory.setKeyDeserializer(StringDeserializer())
        kafkaConsumerFactory.setValueDeserializer(JsonDeserializer<MailMessage>(objectMapper))

        return kafkaConsumerFactory
    }

    private fun buildKafkaConsumerProperties(
        classPathFrom: String,
        classPathTo: String,
        groupId: String
    ): MutableMap<String, Any> {
        val properties: MutableMap<String, Any> = HashMap()
        properties.put(ConsumerConfig.GROUP_ID_CONFIG, groupId)
        properties.put(JsonDeserializer.TYPE_MAPPINGS, String.format("%s:%s", classPathFrom, classPathTo))
        properties.put(JsonDeserializer.TRUSTED_PACKAGES, "*")
        properties.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, kafkaEnvironment.KAFKA_BOOTSTRAP_SERVERS)
        properties.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest")

        return properties
    }
}